### ISSUE TITLE
Feature Request: Make it configurable how a pattern is matched against candidates

### ISSUE THREAD

**Issue by tsdh** [abo-abo/swiper#48]
Right now, swiper/ivy convert the pattern the user has typed to a regexp and use that to filter the candidates.  I like this style for swiper (although I redefine `ivy--regex` as suggested in #47) and buffer completion, and it's ok for function/command/variable completion, but I don't like it at all for file completion.  There, I'd like to have only exact substring matches starting at the first position.  E.g., when I type the pattern `.e` when completing some file name, it should only list hidden files and directories starting with `.e`, e.g., `.emacs` and `.emacs.d/` but not `.TeX/` or `Documents/`.

Long story short: I wish that I could choose a different matching behavior for different kinds of completion just like I can do with the stock emacs completion facility (see `completion-styles` and `completion-category-overrides`).

**Comment by abo-abo** [comment]
So you don't like to use the `^\.e` pattern to match what you need?

I could introduce something like this:

```elisp
(defvar ivy-re-builders-alist
  '((read-file-name-internal . (lambda (x) (concat "^" (regexp-quote x))))
    (t . ivy--regex)))
```

The actual example for `read-file-name-internal` is very limiting, compared to the option to do `^\.e`.
But I suppose that you could write something as intelligent as you want in that place.
Would this work for you?

**Comment by tsdh** [comment]
Typing ^\. myself is too hard for me given that I actually want that limitation always for for completion. Your suggestion fits my bill perfectly.

Hm, when thinking about it, it would be nice if there was a way that lets me bypass pattern mangling altogether (even `ivy--regex`) so that my input pattern is taken as a literal regexp. How about some command in `ivy-minibuffer-map` that toggles pattern mangling off and on again so that I can quickly use it when needed?

**Comment by tsdh** [comment]
Awesome, thank you!
